volttron/platform/install_agents.py

```diff
@@ -0,0 +1,361 @@
+
+    data = execute_command(cmds, env=env, logger=_log,
+                           err_prefix="Error checking identity")
+    for x in data.split("\n"):
+        if x:
+            line_split = x.split()
+            if identity == line_split[2]:
+                return line_split[0]
+    return False
+
+
+def install_requirements(agent_source):
+    req_file = os.path.join(agent_source, "requirements.txt")
+
+    if os.path.exists(req_file):
+        _log.info(f"Installing requirements for agent from {req_file}.")
+        cmds = ["pip", "install", "-r", req_file]
+        try:
+            execute_command(cmds, logger=_log,
+                            err_prefix="Error installing requirements")
+        except RuntimeError:
+            sys.exit(1)
+
+
+def install_agent_directory(opts, package, agent_config):
+    """
+    The main installation method for installing the agent on the correct local
+    platform instance.
+    :param opts:
+    :param package:
+    :param agent_config:
+    :return:
+    """
+    if not os.path.isfile(os.path.join(opts.install_path, "setup.py")):
+        _log.error("Agent source must contain a setup.py file.")
+        sys.exit(-10)
+
+    install_requirements(opts.install_path)
+
+    wheelhouse = os.path.join(get_home(), "packaged")
+    opts.package = create_package(opts.install_path, wheelhouse, opts.vip_identity)
+
+    if not os.path.isfile(opts.package):
+        _log.error("The wheel file for the agent was unable to be created.")
+        sys.exit(-10)
+
+    agent_exists = False
+    volttron_control = os.path.join(get_volttron_root(), "env/bin/vctl")
+    if opts.vip_identity is not None:
+        # if the identity exists the variable will have the agent uuid in it.
+        agent_exists = identity_exists(volttron_control, opts.vip_identity)
+        if agent_exists:
+            if not opts.force:
+                _log.error(
+                    "identity already exists, but force wasn't specified.")
+                sys.exit(-10)
+            # Note we don't remove the agent here because if we do that will
+            # not allow us to update without losing the keys.  The
+            # install_agent method either installs or upgrades the agent.
+
+    if agent_config is None:
+        agent_config = {}
+
+    # if not a dict then config should be a filename
+    if not isinstance(agent_config, dict):
+        config_file = agent_config
+    else:
+        cfg = tempfile.NamedTemporaryFile()
+        with open(cfg.name, 'w') as fout:
+            fout.write(yaml.safe_dump(agent_config))
+        config_file = cfg.name
+
+    try:
+        with open(config_file) as fp:
+            data = yaml.safe_load(fp)
+    except:
+        _log.error("Invalid yaml/json config file.")
+        sys.exit(-10)
+
+    # Configure the whl file before installing.
+    add_files_to_package(opts.package, {'config_file': config_file})
+    env = os.environ.copy()
+
+    if agent_exists:
```

Comment:
We should upgrade only if 'force' is provided

Reply:
This is actually found up above...this if is for upgrade or install.